# To Classify binary variable 'Exited' reflecting the fact whether the customer left the bank (closed his account) or he continues to be a customer.

## Libraries
 Importing numpy, pandas, matplotlib, and seaborn libraries. And then setting %matplotlib inline 


>NumPy is a Linear Algebra Library for Python.Almost all of the libraries in the PyData Ecosystem rely on NumPy.
Numpy is also incredibly fast,as it has bindings to C libraries

>Pandas is an open source library.It is used for fast Analysis,Data Cleaning and Preparation.Using it we can work with wide variety of data

>Matplotlib is a python 2D and 3D plotting library which produces publication quality figures in a variety of hardcopy formats and interactive environments across platforms

>Seaborn is a statistical plotting library with beautiful default styles.it is designed to work with pandas data frame

>%matplotlib inline is used to see plots in the jupyter notebook


In [413]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

## Reading Data

Reading the auto-mpg csv file as a DataFrame called df.

In [414]:
df=pd.read_csv('Churn_Modelling.csv')

In [380]:
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

** This data set contains details of a bank's customers and the target variable is a binary variable reflecting the fact whether the customer left the bank (closed his account) i.e Exited=1 or he continues to be a customer i.e Exited=0
>My aim is to develop a model using Machine Learning Algorithm to Classify the target variable 'Exited' which is a binary variable reflecting the fact whether the customer left the bank (closed his account) or he continues to be a customer
>Data has 6 discrete Attributes i.e RowNumber,CustomerId,CreditScore,Age,Tenure & NumOfProducts , 2 String Categorical Variables i.e Geography & Gender and 2 numeric Categorical Variables i.e HasCrCard & IsActiveMember , 2 Continuous Attributes i.e Balance & EstimatedSalary and a String Attribute as Surname

In [381]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
RowNumber          10000 non-null int64
CustomerId         10000 non-null int64
Surname            10000 non-null object
CreditScore        10000 non-null int64
Geography          10000 non-null object
Gender             10000 non-null object
Age                10000 non-null int64
Tenure             10000 non-null int64
Balance            10000 non-null float64
NumOfProducts      10000 non-null int64
HasCrCard          10000 non-null int64
IsActiveMember     10000 non-null int64
EstimatedSalary    10000 non-null float64
Exited             10000 non-null int64
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


** Pandas dataframe.info() function is used to get a concise summary of the dataframe

>** From info() we come to know that total entries in the data frame is 10000.Given data doesn't contain any missing observation. Also the attribute Geography,Gender & Surname are of type object and rest of the attributes contain 
numeric values

In [411]:
df.describe()

CreditScore        Gender           Age        Tenure        Balance  \
count  10000.000000  10000.000000  10000.000000  10000.000000   10000.000000   
mean     650.528800      0.545700     38.921800      5.012800   76485.889288   
std       96.653299      0.497932     10.487806      2.892174   62397.405202   
min      350.000000      0.000000     18.000000      0.000000       0.000000   
25%      584.000000      0.000000     32.000000      3.000000       0.000000   
50%      652.000000      1.000000     37.000000      5.000000   97198.540000   
75%      718.000000      1.000000     44.000000      7.000000  127644.240000   
max      850.000000      1.000000     92.000000     10.000000  250898.090000   

       NumOfProducts    HasCrCard  IsActiveMember  EstimatedSalary  \
count   10000.000000  10000.00000    10000.000000     10000.000000   
mean        1.530200      0.70550        0.515100    100090.239881   
std         0.581654      0.45584        0.499797     57510.492818   
min         1.000000      0.00000        0.000000        11.580000   
25%         1.000000      0.00000        0.000000     51002.110000   
50%         1.000000      1.00000        1.000000    100193.915000   
75%         2.000000      1.00000        1.000000    149388.247500   
max         4.000000      1.00000        1.000000    199992.480000   

             Exited       Germany         Spain  
count  10000.000000  10000.000000  10000.000000  
mean       0.203700      0.250900      0.247700  
std        0.402769      0.433553      0.431698  
min        0.000000      0.000000      0.000000  
25%        0.000000      0.000000      0.000000  
50%        0.000000      0.000000      0.000000  
75%        0.000000      1.000000      0.000000  
max        1.000000      1.000000      1.000000

** Pandas describe() is used to view some basic statistical details like percentile, mean, std etc. of a data frame or a series of numeric values.

## Exploratory Data Analysis

In [383]:
%%HTML
<script type='text/javascript' src='https://eu-west-1a.online.tableau.com/javascripts/api/viz_v1.js'></script><div class='tableauPlaceholder' style='width: 1000px; height: 827px;'><object class='tableauViz' width='1000' height='827' style='display:none;'><param name='host_url' value='https%3A%2F%2Feu-west-1a.online.tableau.com%2F' /> <param name='embed_code_version' value='3' /> <param name='site_root' value='&#47;t&#47;datascienceproject' /><param name='name' value='BankProject&#47;DashboardBankProject' /><param name='tabs' value='no' /><param name='toolbar' value='yes' /><param name='showAppBanner' value='false' /><param name='filter' value='iframeSizedToWindow=true' /></object></div>

# Data Cleaning

Attributes 'Surname','RowNumber','CustomerId' are not relevant for 'Data Modeling' but it is relevant for 'Data Analysis' .So, we should remove it from the data frame.

In [384]:
df.drop(['Surname','RowNumber','CustomerId'],axis=1,inplace=True)

In [385]:
df.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42       2       0.00              1   
1          608     Spain  Female   41       1   83807.86              1   
2          502    France  Female   42       8  159660.80              3   
3          699    France  Female   39       1       0.00              2   
4          850     Spain  Female   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1        101348.88       1  
1          0               1        112542.58       0  
2          1               0        113931.57       1  
3          0               0         93826.63       0  
4          1               1         79084.10       0

We'll need to convert categorical features to dummy variables , Otherwise our machine learning algorithm won't be able to directly take in those features as inputs. Columns Geography,Gender,HasCrCard	& IsActiveMember are of categorical type.But columns HasCrCard & IsActiveMember are in proper categorical format so we needn't to add dummy variables for these attributes.  we will add Dummy variables only for Geography,Gender attributes.

>pd.get_dummies() :function which can turn a categorical variable into a series of zeros and ones,which makes them a lot
easier to quantify and compare
>>drop_first=True : Since one of the columns can be generated completely from the others, and hence retaining this extra 
column does not add any new information for the modelling process, it will be good practice to always drop the first column 

>pd.concat() :function which can perform concatenation operations along an axis

>dataframe.drop() :Pandas drop function allows us to drop/remove one or more columns from a dataframe.
>>inplace=True :this operation changes directly the content of a given data without making a copy.

In [386]:
from sklearn.preprocessing import LabelEncoder

In [387]:
df.Gender=LabelEncoder().fit_transform(df.Gender)

In [388]:
df.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France       0   42       2       0.00              1   
1          608     Spain       0   41       1   83807.86              1   
2          502    France       0   42       8  159660.80              3   
3          699    France       0   39       1       0.00              2   
4          850     Spain       0   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1        101348.88       1  
1          0               1        112542.58       0  
2          1               0        113931.57       1  
3          0               0         93826.63       0  
4          1               1         79084.10       0

In [389]:
Geography1=pd.get_dummies(df['Geography'],drop_first=True)

In [390]:
df=pd.concat([df,Geography1],axis=1)

In [391]:
df.drop(['Geography'],axis=1,inplace=True)

In [392]:
df.head()

CreditScore  Gender  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          619       0   42       2       0.00              1          1   
1          608       0   41       1   83807.86              1          0   
2          502       0   42       8  159660.80              3          1   
3          699       0   39       1       0.00              2          0   
4          850       0   43       2  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Exited  Germany  Spain  
0               1        101348.88       1        0      0  
1               1        112542.58       0        0      1  
2               0        113931.57       1        0      0  
3               0         93826.63       0        0      0  
4               1         79084.10       0        0      1

StandardScaler performs the task of Standardization. Usually a dataset contains variables that are different in scale.
>It will transform the data such that its distribution will have a mean value 0 and standard deviation of 1. Given the 
distribution of the data, each value in the dataset will have the sample mean value subtracted, and then divided by the
standard deviation of the whole dataset.

In [393]:
from sklearn.preprocessing import StandardScaler

In [394]:
sc=StandardScaler()

** Transforming the data into standard scale

We can perform two steps fitting and transforming dataset in single step using "fit_transform " 
>fit method is applied to the dataset to learns the model parameters (for example, mean and standard deviation).
>transform method is apllied to the dataset to get the transformed (scaled) dataset.

In [395]:
sc.fit_transform(df)

C:\Users\shiv\Anaconda3\lib\site-packages\sklearn\preprocessing\data.py:645: DataConversionWarning: Data with input dtype uint8, int32, int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
C:\Users\shiv\Anaconda3\lib\site-packages\sklearn\base.py:464: DataConversionWarning: Data with input dtype uint8, int32, int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


array([[-0.32622142, -1.09598752,  0.29351742, ...,  1.97716468,
        -0.57873591, -0.57380915],
       [-0.44003595, -1.09598752,  0.19816383, ..., -0.50577476,
        -0.57873591,  1.74273971],
       [-1.53679418, -1.09598752,  0.29351742, ...,  1.97716468,
        -0.57873591, -0.57380915],
       ...,
       [ 0.60498839, -1.09598752, -0.27860412, ...,  1.97716468,
        -0.57873591, -0.57380915],
       [ 1.25683526,  0.91241915,  0.29351742, ...,  1.97716468,
         1.72790383, -0.57380915],
       [ 1.46377078, -1.09598752, -1.04143285, ..., -0.50577476,
        -0.57873591, -0.57380915]])

In [396]:
df.head()

CreditScore  Gender  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          619       0   42       2       0.00              1          1   
1          608       0   41       1   83807.86              1          0   
2          502       0   42       8  159660.80              3          1   
3          699       0   39       1       0.00              2          0   
4          850       0   43       2  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Exited  Germany  Spain  
0               1        101348.88       1        0      0  
1               1        112542.58       0        0      1  
2               0        113931.57       1        0      0  
3               0         93826.63       0        0      0  
4               1         79084.10       0        0      1

## Train Test Split
Using model_selection.train_test_split from sklearn to split the data into training and testing sets.

In [397]:
from sklearn.model_selection import train_test_split

scikit-learn provides a helpful function for partitioning data, train_test_split ,which splits out our data into a training set
and a test set
>test_size : float, int or None, optional (default=None)
>>If float, should be between 0.0 and 1.0 and represent the proportion of the dataset to include in the test split.
>>If int, represents the absolute number of test samples. 
>>If None, the value is set to the complement of the train size. If train_size is also None, it will be set to 0.25.

>train_size : float, int, or None, (default=None)
>>If float, should be between 0.0 and 1.0 and represent the proportion of the dataset to include in the train split. 
>>If int, represents the absolute number of train samples. 
>>If None, the value is automatically set to the complement of the test size.

>random_state : int, RandomState instance or None, optional(default=None)
>>If int, random_state is the seed used by the random number generator
>>If RandomState instance, random_state is the random number generator
>>If None, the random number generator is the RandomState instance used by np.random.

>shuffle : boolean, optional (default=True)
>>Whether or not to shuffle the data before splitting. If shuffle=False then stratify must be None.

>stratify : array-like or None (default=None)
>>If not None, data is split in a stratified fashion, using this as the class labels.


## Training and Testing Data

In [398]:
X = df.drop(['Exited'],axis=1)
y = df['Exited']

In [399]:
X=sc.fit_transform(X)
X

C:\Users\shiv\Anaconda3\lib\site-packages\sklearn\preprocessing\data.py:645: DataConversionWarning: Data with input dtype uint8, int32, int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
C:\Users\shiv\Anaconda3\lib\site-packages\sklearn\base.py:464: DataConversionWarning: Data with input dtype uint8, int32, int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


array([[-0.32622142, -1.09598752,  0.29351742, ...,  0.02188649,
        -0.57873591, -0.57380915],
       [-0.44003595, -1.09598752,  0.19816383, ...,  0.21653375,
        -0.57873591,  1.74273971],
       [-1.53679418, -1.09598752,  0.29351742, ...,  0.2406869 ,
        -0.57873591, -0.57380915],
       ...,
       [ 0.60498839, -1.09598752, -0.27860412, ..., -1.00864308,
        -0.57873591, -0.57380915],
       [ 1.25683526,  0.91241915,  0.29351742, ..., -0.12523071,
         1.72790383, -0.57380915],
       [ 1.46377078, -1.09598752, -1.04143285, ..., -1.07636976,
        -0.57873591, -0.57380915]])

In [400]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

## Training the Model
Logistic Regression is one of the basic and popular algorithm to solve a classification problem. The logistic function is a Sigmoid function, which takes any real value between zero and one.And if we plot it, the graph will be S curve.

> Logistic regression is the appropriate regression analysis to conduct when the dependent variable is dichotomous (binary).  Like all regression analyses, the logistic regression is a predictive analysis.Logistic regression is used to describe data and to explain the relationship between one dependent binary variable and one or more nominal, ordinal, interval or ratio-level independent variables

In [401]:
from sklearn.linear_model import LogisticRegression

Creating an instance of a LogisticRegression() model named as logmodel and Fitting the training data set

In [402]:
logmodel = LogisticRegression()
logmodel.fit(X_train,y_train)

C:\Users\shiv\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

## Predicting Test Data

In [403]:
predictions = logmodel.predict(X_test)

## Evaluating Model

evaluate the model by checking its classification report

**The classification report shows a representation of the main classification metrics on a per-class basis. This gives a deeper intuition of the classifier behavior over global accuracy which can mask functional weaknesses in one class of a multiclass problem.

The metrics are defined in terms of true and false positives, and true and false negatives. Positive and negative classes of a binary classification problem.Therefore a true positive is when the actual class is positive as is the estimated class. A false positive is when the actual class is negative but the estimated class is positive.

>precision : Precision is the ability of a classiifer not to label an instance positive that is actually negative. For each class it is defined as as the ratio of true positives to the sum of true and false positives. Said another way, “for all instances classified positive, what percent was correct?”

>recall : Recall is the ability of a classifier to find all positive instances. For each class it is defined as the ratio of true positives to the sum of true positives and false negatives. Said another way, “for all instances that were actually positive, what percent was classified correctly?”

>f1 score : The F1 score is a weighted harmonic mean of precision and recall such that the best score is 1.0 and the worst is 0.0. Generally speaking, F1 scores are lower than accuracy measures as they embed precision and recall into their computation. As a rule of thumb, the weighted average of F1 should be used to compare classifier models, not global accuracy.

>support : Support is the number of actual occurrences of the class in the specified dataset. Imbalanced support in the training data may indicate structural weaknesses in the reported scores of the classifier and could indicate the need for stratified sampling or rebalancing. Support doesn’t change between models but instead diagnoses the evaluation process.

>>macro average : averaging the unweighted mean per label. A macro-average will compute the metric independently for each class and then take the average (hence treating all classes equally)

>>micro average : averaging the total true positives, false negatives and false positives. it is only shown for multi-label or multi-class with a subset of classes because it is accuracy otherwise. a micro-average will aggregate the contributions of all classes to compute the average metric. In a multi-class classification setup, micro-average is preferable if you suspect there might be class imbalance (i.e you may have many more examples of one class than of other classes).

>>weighted average : averaging the support-weighted mean per label

>>sample average : only for multilabel classification



In [404]:
from sklearn.metrics import classification_report

In [405]:
print(classification_report(y_test,predictions))

              precision    recall  f1-score   support

           0       0.83      0.96      0.89      2657
           1       0.54      0.21      0.30       643

   micro avg       0.81      0.81      0.81      3300
   macro avg       0.69      0.58      0.60      3300
weighted avg       0.78      0.81      0.78      3300



## Conclusion

After Analyzing the Churn Modelling classification data set, we can conclude that the model that we have created using Logistic Regression is 81 % Accurate.That means the model does a very good job in predicting whether the customer left the bank (closed his account) or he continues to be a customer.Small number of observations predict that there will customer left the bank (closed his account). Majority of observations predict that there will continues to be a customer.